### PR TITLE
Add train code for Faster RCNN

### PR DESCRIPTION
**Pull request by yuyu2172**
Files added

+ `links/model/faster_rcnn/utils/anchor_target_creator.py`
+ `links/model/faster_rcnn/utils/proposal_target_creator.py`
+ `links/model/faster_rcnn/faster_rcnn_loss.py`

### CODE REVIEW

**Comment by yuyu2172** [comment]
# Performance

### Evaluated on  VOC 2007 Test

Training protocols are same across implementations.
+ end2end
+ 50000 iterations with `lr=1e-3`, 20000 iterations with `lr=1e-4

Caffe version:   69.9 mAP
Chainer version:  70.5 mAP

Since there are stochastic samplings, performance vary from different random seeds used during *training*. Performance is deterministic in test time.

**Comment by yuyu2172** [comment]
@Hakuyume 
Can you check this PR?

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+    .. [#] Shaoqing Ren, Kaiming He, Ross Girshick, Jian Sun. \
+    Faster R-CNN: Towards Real-Time Object Detection with \
+    Region Proposal Networks. NIPS 2015.
+
+    Args:
+        faster_rcnn (~chainercv.links.model.faster_rcnn.FasterRCNN):
+            A Faster R-CNN model to train.
+        rpn_sigma (float): Sigma parameter for localization loss
+            of RPN.
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: It is better to describe the default value.
`RPN` -> `Region Proposal Network (RPN)`

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        sigma (float): Sigma paramter for localization loss of
+            calculated from the output of the head.
+        anchor_target_creator_params (dict): Key valued paramters for
+            :obj:`chainercv.links.model.faster_rcnn.AnchorTargetCreator`.
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: Why don't you take an instance of `AnchorTargetCreator` directly? I think it is more simple to pass  faster_rcnn_params` as an argument of `AnchorTargetCreator.__call__`.

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        Here are notations used.
+
+        * :math:`N` is the number of batch size.
+        * :math:`R` is the number of bounding boxes per image.
+
+        Currently only :math:`N=1` is supported.
+
+
+        Args:
+            imgs (~chainer.Variable): 4D image variable.
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: `4D image` is `(N, C, H, W)`? If so, `A batch of images` is better.

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        if isinstance(labels, chainer.Variable):
+            labels = labels.data
+        if isinstance(scale, chainer.Variable):
+            scale = scale.data
+        scale = np.asscalar(cuda.to_cpu(scale))
+        n = bboxes.shape[0]
+        if n != 1:
+            raise ValueError('currently only batch size 1 is supported')
+
+        img_size = imgs.shape[2:][::-1]
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]:
```
_, _, H, W = imgs.shape
img_size = (W, H)
```

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        chainer.reporter.report({'rpn_loc_loss': rpn_loc_loss,
+                                 'rpn_cls_loss': rpn_cls_loss,
+                                 'loc_loss': loc_loss,
+                                 'cls_loss': cls_loss,
+                                 'loss': loss},
+                                self)
+        return loss
+
+
+def _smooth_l1_loss(x, t, in_weight, out_weight, sigma):
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: Can you use `chainer.functions.huber_loss` in this function?

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+import chainer
+from chainer import cuda
+
+from chainercv.links.model.faster_rcnn.utils.bbox2loc import bbox2loc
+from chainercv.utils.bbox.bbox_iou import bbox_iou
+
+
+class AnchorTargetCreator(object):
+
+    """Assign anchors to the ground-truth targets.
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: `ground-truth` -> `ground truth`

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    Bounding regression targets are computed using encoding scheme
+    found in :obj:`chainercv.links.model.faster_rcnn.bbox2loc`.
+
+    .. [#] Shaoqing Ren, Kaiming He, Ross Girshick, Jian Sun. \
+    Faster R-CNN: Towards Real-Time Object Detection with \
+    Region Proposal Networks. NIPS 2015.
+
+    Args:
+        n_sample (int): Number of regions to produce.
+        positive_iou (float): Anchors with IoU above this
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: How about `positive_iou_thresh`?

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+            threshold will be assigned as positive.
+        negative_iou (float): Anchors with IoU below this
+            threshold will be assigned as negative.
+        fg_fraction (float): Fraction of positive regions in the
+            set of all regions produced.
+        loc_in_weight (tuple of four floats): Four coefficients
+            used to calculate loc_in_weight.
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: `loc_in_weight` is used to calculate `loc_in_weight`?

Comment (yuyu2172) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: The original name was `loc_inside_weight`.
Perhaps, this was confusing.
What do you think about `loc_weight_in`?

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py (hunk 2 above)
Comment (yuyu2172) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: OK.
That makes sense.

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py (hunk 8 above)
Comment (yuyu2172) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: I misunderstood what you said.
How about `loc_weight_in_base` is used as a parameter, and `loc_weight_in` is returned?

**Comment by yuyu2172** [comment]
Thanks for reviewing.

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py (hunk 8 above)
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: How about using `loc_weight_in_coeff` for the input?

Comment (yuyu2172) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: Isn't `weight` and `coeff` same thing?

Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: Sorry, I misunderstood. `loc_in_weight_base` and `loc_in_weight` look good to me.

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py (hunk 5 above)
Comment (yuyu2172) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: Thanks for your suggestion.
I looked into it, and I found that parameterization used in the Huber loss and smooth L1 are different.

I think that the current style is OK.
After merging this, I am planning to add `_bbox_loss` to `links.loss` with a proper documentation on the formula used.

**Comment by yuyu2172** [comment]
I changed `*TargetCreator` to stop returning `*_weight`.
I think the functionality of `*TargetCreator` became more explicit by the change.

**Comment by yuyu2172** [comment]
This is ready for review.

**Comment by Hakuyume** [comment]
OK, I will check this PR.

File: chainercv/links/__init__.py [hunk 9]
```diff
@@ -1,5 +1,6 @@
 from chainercv.links.model.pixelwise_softmax_classifier import PixelwiseSoftmaxClassifier  # NOQA
 
+from chainercv.links.model.faster_rcnn.faster_rcnn_train_chain import FasterRCNNTrainChain  # NOQA
```
Comment (Hakuyume) [inline chainercv/links/__init__.py]: This import is against to the rule that we only import models which have APIs of detection or semantic segmentation.
I know `PixelwiseSoftmaxClassifier` is against too. Perhaps, we have to change the rule. Do you have any ideas?

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+    .. [#] Shaoqing Ren, Kaiming He, Ross Girshick, Jian Sun. \
+    Faster R-CNN: Towards Real-Time Object Detection with \
+    Region Proposal Networks. NIPS 2015.
+
+    Args:
+        faster_rcnn (~chainercv.links.model.faster_rcnn.FasterRCNN):
+            A Faster R-CNN model that is going to be trained.
+        rpn_sigma (float): Sigma parameter for the localization loss
+            of Region Proposal Network (RPN).
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: How about adding a note about the default value? Is it the value used in the original paper?

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
+import numpy as np
+
+import chainer
+from chainer import cuda
+import chainer.functions as F
+
+from chainercv.links.model.faster_rcnn.utils.anchor_target_creator import \
+    AnchorTargetCreator
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: How about `from chainercv.links.model.faster_rcnn import AnchorTargetCreator`?

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        roi_cls_loc = roi_cls_loc.reshape(n_sample, -1, 4)
+        roi_loc = roi_cls_loc[self.xp.arange(n_sample), gt_roi_label]
+        roi_loc_loss = _fast_rcnn_loc_loss(
+            roi_loc, gt_roi_loc, gt_roi_label, self.sigma)
+        roi_cls_loss = F.softmax_cross_entropy(roi_score, gt_roi_label)
+
+        loss = rpn_loc_loss + rpn_cls_loss + roi_loc_loss + roi_cls_loss
+        chainer.reporter.report({'rpn_loc_loss': rpn_loc_loss,
+                                 'rpn_cls_loss': rpn_cls_loss,
+                                 'roi_loc_loss': roi_loc_loss,
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: Is `roi_loc_loss` a loss from the head module?
According to the docs, the losses are computed form PRN and the head.
> This is the sum of losses for Region Proposal Network and the head module.

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+from chainercv.links.model.faster_rcnn.utils.bbox2loc import bbox2loc
+from chainercv.utils.bbox.bbox_iou import bbox_iou
+
+
+class AnchorTargetCreator(object):
+
+    """Assign the ground truth bounding boxes to anchors.
+
+    Assigns the ground-truth bounding boxes to anchors for training Region
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: `ground-truth` -> `ground truth`

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    Region Proposal Networks. NIPS 2015.
+
+    Args:
+        n_sample (int): The number of regions to produce.
+        positive_iou_thresh (float): Anchors with IoU above this
+            threshold will be assigned as positive.
+        negative_iou_thresh (float): Anchors with IoU below this
+            threshold will be assigned as negative.
+        fg_fraction (float): Fraction of positive regions in the
+            sampled regions.
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: How about `positive_ratio`?

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+
+        Types of input arrays and output arrays are same.
+
+        Here are notations.
+
+        * :math:`S` is the number of anchors.
+        * :math:`R` is the number of bounding boxes.
+
+        Args:
+            bbox (array): Coodinates of bounding boxes. Its shape is
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: `Coodinates` -> `Coordinates`

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py (hunk 14 above)
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: In this function, you are using both `positive/negative` and `foreground/background`. Are they different from each other?

File: chainercv/links/model/faster_rcnn/utils/proposal_target_creator.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+                is :math:`(R',)`.
+            loc_normalize_mean (tuple of four floats): Mean values to normalize
+                coordinates of bouding boxes.
+            loc_normalize_std (tupler of four floats): Standard deviation of
+                the coordinates of bounding boxes.
+
+        Returns:
+            (array, array, array):
+
+            * **sample_roi**: Regions of interests that are sampled. \
```
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/utils/proposal_target_creator.py]: For the consistency with `IoU`, `Regions of interests` -> `Regions of Interests`?
How about `Sampled Regions of Interests (RoIs).`?

File: examples/faster_rcnn/train_faster_rcnn.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
+from __future__ import division
```
Comment (Hakuyume) [inline examples/faster_rcnn/train_faster_rcnn.py]: I think `train.py` is enough because this is a directory for `Faster R-CNNc`.

File: tests/links_tests/model_tests/faster_rcnn_tests/test_faster_rcnn.py [hunk 18]
```diff
@@ -174,6 +175,44 @@ def test_predict_gpu(self):
         self.check_predict()
 
 
+class TestFasterRCNNTrainChain(unittest.TestCase):
```
Comment (Hakuyume) [inline tests/links_tests/model_tests/faster_rcnn_tests/test_faster_rcnn.py]: I think this test should be separated into another file.

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py (hunk 12 above)
Comment (yuyu2172) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: Yes. This is a loss computed using `roi_*`.

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py (hunk 11 above)
Comment (yuyu2172) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: That produces import error when I run tests.

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py (hunk 12 above)
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: I see. How about documenting it?
For example,
```
This is the sum of losses for Region Proposal Network (:obj:`rpn_loc_loss`/:obj:`rpn_cls_loss`)  and the head module (:obj:`roi_loc_loss`/:obj:`roi_cls_loss`).
```

File: chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py (hunk 11 above)
Comment (Hakuyume) [inline chainercv/links/model/faster_rcnn/faster_rcnn_train_chain.py]: I see.

File: chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py (hunk 14 above)
Comment (yuyu2172) [inline chainercv/links/model/faster_rcnn/utils/anchor_target_creator.py]: Since `positive` is quite long, how about `pos_ratio`?

**Comment by yuyu2172** [comment]
Fixed the code. Thanks for reviewing.

File: examples/faster_rcnn/README.md [hunk 19]
```diff
@@ -41,6 +41,15 @@ The integer convention encodes right below vertex coordinates of bounding boxes
 On top of that, the anchors are not discretized in ChainerCV.
 
 
+### Train code
+
+You can train the model with the following code.
+
+```
+$ python train.py [--gpu <gpu]
```
Comment (Hakuyume) [inline examples/faster_rcnn/README.md]: `[--gpu <gpu]` -> `[--gpu <gpu>]`

**Comment by Hakuyume** [comment]
LGTM
